File: ios/SuperCameraKit/Sources/SuperCameraKit/SuperCamera.swift
import Foundation

public enum WorkMode { case purePreview, analysis, photo, video }
public enum CameraState: Equatable { case idle, initializing, previewing, capturing, recording, releasing }
public enum SCErrorCode: Equatable {
    case none
    case permission
    case deviceUnavailable
    case sessionConfigFailed
    case renderInitFailed
    case renderSurfaceLost
    case encoderInitFailed
    case encoderBackpressure
    case fileIOFailed
    case thermalThrottle
    case invalidState
}

public typealias SCErrorCallback = (SCErrorCode, String) -> Void

public final class SuperCamera {
    private var mode: WorkMode = .purePreview
    private var state: CameraState = .idle
    private var errorCallback: SCErrorCallback?
    private var recordPath: String = ""
    private let previewSession: PreviewSessionPort
    private let previewBridge: PreviewBridgePort
    private let photoCapture: PhotoCapturePort

    public init(
        previewSession: PreviewSessionPort = NoOpPreviewSessionPort(),
        previewBridge: PreviewBridgePort = PreviewRenderBridge(),
        photoCapture: PhotoCapturePort = CaptureControllerAdapter()
    ) {
        self.previewSession = previewSession
        self.previewBridge = previewBridge
        self.photoCapture = photoCapture
    }

    public func setWorkMode(_ mode: WorkMode) { self.mode = mode }
    public func currentWorkMode() -> WorkMode { mode }
    public func currentState() -> CameraState { state }
    public func setErrorCallback(_ cb: @escaping SCErrorCallback) { errorCallback = cb }

    @discardableResult
    public func startPreview(surfaceHandle: UInt64) -> Bool {
        guard state == .idle else {
            errorCallback?(.invalidState, "startPreview requires idle state, current=\(state)")
            return false
        }
        state = .initializing
        guard previewSession.startPreview(surfaceHandle: surfaceHandle) else {
            state = .idle
            errorCallback?(.sessionConfigFailed, "preview session init failed")
            return false
        }
        guard previewBridge.attach(surfaceHandle: surfaceHandle) else {
            _ = previewSession.stopPreview()
            state = .idle
            errorCallback?(.renderInitFailed, "preview bridge attach failed")
            return false
        }
        state = .previewing
        return true
    }

    @discardableResult
    public func stopPreview() -> Bool {
        guard state == .previewing else {
            errorCallback?(.invalidState, "stopPreview requires previewing state, current=\(state)")
            return false
        }
        state = .releasing
        previewBridge.detach()
        let stopped = previewSession.stopPreview()
        state = .idle
        if !stopped {
            errorCallback?(.sessionConfigFailed, "preview session stop failed")
            return false
        }
        return true
    }

    @discardableResult
    public func startRecord(path: String) -> Bool {
        guard state == .previewing else {
            errorCallback?(.invalidState, "startRecord requires previewing state, current=\(state)")
            return false
        }
        recordPath = path
        state = .recording
        return true
    }

    public func stopRecord() -> RecordResult {
        guard state == .recording else {
            errorCallback?(.invalidState, "stopRecord requires recording state, current=\(state)")
            return RecordResult(path: "")
        }
        state = .previewing
        return RecordResult(path: recordPath)
    }

    public func takePhoto(outputPath: String) -> CaptureResult {
        guard state == .previewing else {
            errorCallback?(.invalidState, "takePhoto requires previewing state, current=\(state)")
            return CaptureResult(path: "")
        }
        state = .capturing
        let result = photoCapture.capture(outputPath: outputPath)
        state = .previewing
        if result.path.isEmpty {
            errorCallback?(.fileIOFailed, "photo capture returned empty path")
        }
        return result
    }
}

File: android/supercamera/src/main/java/com/dcamera/SuperCamera.kt
package com.dcamera

import com.dcamera.capture.CaptureResult
import com.dcamera.core.CaptureControllerAdapter
import com.dcamera.core.ErrorCallback
import com.dcamera.core.NoOpPreviewSessionPort
import com.dcamera.core.PhotoCapturePort
import com.dcamera.core.PreviewSessionPort
import com.dcamera.record.RecordResult
import com.dcamera.render.PreviewBridgePort
import com.dcamera.render.PreviewRenderBridge

enum class WorkMode { PURE_PREVIEW, ANALYSIS, PHOTO, VIDEO }
enum class CameraState { IDLE, INITIALIZING, PREVIEWING, CAPTURING, RECORDING, RELEASING }
enum class ErrorCode {
    NONE,
    PERMISSION,
    DEVICE_UNAVAILABLE,
    SESSION_CONFIG_FAILED,
    RENDER_INIT_FAILED,
    RENDER_SURFACE_LOST,
    ENCODER_INIT_FAILED,
    ENCODER_BACKPRESSURE,
    FILE_IO_FAILED,
    THERMAL_THROTTLE,
    INVALID_STATE
}

class SuperCamera(
    private val previewSession: PreviewSessionPort = NoOpPreviewSessionPort(),
    private val previewBridge: PreviewBridgePort = PreviewRenderBridge(),
    private val photoCapture: PhotoCapturePort = CaptureControllerAdapter(),
) {
    private var mode: WorkMode = WorkMode.PURE_PREVIEW
    private var state: CameraState = CameraState.IDLE
    private var errorCallback: ErrorCallback? = null
    private var recordPath: String = ""

    fun setWorkMode(next: WorkMode) { mode = next }
    fun currentWorkMode(): WorkMode = mode
    fun currentState(): CameraState = state
    fun setErrorCallback(cb: ErrorCallback) { errorCallback = cb }

    fun startPreview(surfaceHandle: Long): Boolean {
        if (state != CameraState.IDLE) {
            errorCallback?.onError(
                ErrorCode.INVALID_STATE,
                "startPreview requires IDLE state, current=$state",
            )
            return false
        }
        state = CameraState.INITIALIZING
        previewSession.setFrameConsumer { previewBridge.submitFrame(it) }
        if (!previewSession.startPreview(surfaceHandle)) {
            state = CameraState.IDLE
            errorCallback?.onError(ErrorCode.SESSION_CONFIG_FAILED, "preview session init failed")
            return false
        }
        if (!previewBridge.attachSurface(surfaceHandle)) {
            previewSession.stopPreview()
            state = CameraState.IDLE
            errorCallback?.onError(ErrorCode.RENDER_INIT_FAILED, "preview bridge attach failed")
            return false
        }
        state = CameraState.PREVIEWING
        return true
    }

    fun stopPreview(): Boolean {
        if (state != CameraState.PREVIEWING) {
            errorCallback?.onError(
                ErrorCode.INVALID_STATE,
                "stopPreview requires PREVIEWING state, current=$state",
            )
            return false
        }
        state = CameraState.RELEASING
        previewBridge.detachSurface()
        val stopped = previewSession.stopPreview()
        state = CameraState.IDLE
        if (!stopped) {
            errorCallback?.onError(ErrorCode.SESSION_CONFIG_FAILED, "preview session stop failed")
            return false
        }
        return true
    }

    fun startRecord(path: String): Boolean {
        if (state != CameraState.PREVIEWING) {
            errorCallback?.onError(
                ErrorCode.INVALID_STATE,
                "startRecord requires PREVIEWING state, current=$state",
            )
            return false
        }
        recordPath = path
        state = CameraState.RECORDING
        return true
    }

    fun stopRecord(): RecordResult {
        if (state != CameraState.RECORDING) {
            errorCallback?.onError(
                ErrorCode.INVALID_STATE,
                "stopRecord requires RECORDING state, current=$state",
            )
            return RecordResult("")
        }
        state = CameraState.PREVIEWING
        return RecordResult(recordPath)
    }

    fun takePhoto(outputPath: String): CaptureResult {
        if (state != CameraState.PREVIEWING) {
            errorCallback?.onError(
                ErrorCode.INVALID_STATE,
                "takePhoto requires PREVIEWING state, current=$state",
            )
            return CaptureResult("")
        }
        state = CameraState.CAPTURING
        val result = photoCapture.capture(outputPath)
        state = CameraState.PREVIEWING
        if (result.path.isBlank()) {
            errorCallback?.onError(ErrorCode.FILE_IO_FAILED, "photo capture returned empty path")
        }
        return result
    }
}
